File: src/agent_nexus/models/evolution.py
"""Self-Evolution Engine models: SkillRecord, EvolutionType, SkillOrigin, SkillLineage, EvolutionContext, EvolutionMetrics."""  # noqa: E501

from __future__ import annotations

from datetime import datetime
from enum import StrEnum

from pydantic import Field, model_validator

from agent_nexus.models._common import FrozenModel, _utc_now


def _validate_counter_invariants(
    total_selections: int,
    total_applied: int,
    total_completions: int,
    total_fallbacks: int,
) -> None:
    """Shared counter invariant validation for SkillRecord and EvolutionMetrics."""
    if total_selections == 0 and (total_applied != 0 or total_fallbacks != 0):
        raise ValueError(
            "counter invariant violated: zero selections requires zero applied and zero fallbacks"
        )
    if total_applied > total_selections:
        raise ValueError("total_applied cannot exceed total_selections")
    if total_completions > total_applied:
        raise ValueError("total_completions cannot exceed total_applied")
    if total_fallbacks > total_applied:
        raise ValueError("total_fallbacks cannot exceed total_applied")
    if total_completions + total_fallbacks > total_applied:
        raise ValueError("total_completions + total_fallbacks cannot exceed total_applied")


class EvolutionType(StrEnum):
    """How a Skill was produced by the Evolution Engine.

    - FIX: A broken Skill was repaired (1 parent).
    - DERIVED: A new Skill derived from 1+ existing Skills (enhancement/merge).
    - CAPTURED: A new Skill extracted from a successful task with no parent Skill (0 parents).
    """

    FIX = "fix"
    DERIVED = "derived"
    CAPTURED = "captured"


class SkillOrigin(StrEnum):
    """Where a Skill originally came from.

    - IMPORTED: Shipped with the Agent Package.
    - CAPTURED: Created by the Evolution Engine from a successful task.
    - DERIVED: Created by enhancing/merging existing Skills.
    - FIXED: Created by repairing a broken Skill.
    """

    IMPORTED = "imported"
    CAPTURED = "captured"
    DERIVED = "derived"
    FIXED = "fixed"


class SkillLineage(FrozenModel):
    """DAG version history of a Skill.

    Tracks the origin, generation number, parent Skills (for FIX/DERIVED),
    content diff/snapshot for auditing.
    """

    origin: SkillOrigin = SkillOrigin.IMPORTED
    generation: int = Field(default=0, ge=0)
    parent_skill_ids: list[str] = Field(default_factory=list)
    content_diff: str | None = None
    content_snapshot: dict[str, str] | None = None


class SkillRecord(FrozenModel):
    """A Skill in the Evolution Engine's registry.

    Each SkillRecord tracks quality counters used by health diagnostics
    and evolution triggers:
        applied_rate      = total_applied / total_selections
        completion_rate   = total_completions / total_applied
        effective_rate    = total_completions / total_selections
        fallback_rate     = total_fallbacks / total_selections
    """

    id: str = Field(min_length=1)
    name: str = Field(min_length=1)
    version: str = "1.0.0"
    lineage: SkillLineage = Field(default_factory=SkillLineage)
    directory: str = ""
    is_active: bool = True
    total_selections: int = Field(default=0, ge=0)
    total_applied: int = Field(default=0, ge=0)
    total_completions: int = Field(default=0, ge=0)
    total_fallbacks: int = Field(default=0, ge=0)
    first_seen: datetime = Field(default_factory=_utc_now)
    last_updated: datetime = Field(default_factory=_utc_now)

    @model_validator(mode="after")
    def _validate_counters(self) -> SkillRecord:
        _validate_counter_invariants(
            self.total_selections, self.total_applied,
            self.total_completions, self.total_fallbacks,
        )
        return self


class EvolutionMetrics(FrozenModel):
    """Aggregated evolution quality metrics for an Agent.

    Used in L0 context injection (~30 tokens).
    """

    total_selections: int = Field(default=0, ge=0)
    total_applied: int = Field(default=0, ge=0)
    total_completions: int = Field(default=0, ge=0)
    total_fallbacks: int = Field(default=0, ge=0)

    @model_validator(mode="after")
    def _validate_counters(self) -> EvolutionMetrics:
        _validate_counter_invariants(
            self.total_selections, self.total_applied,
            self.total_completions, self.total_fallbacks,
        )
        return self


class EvolutionContext(FrozenModel):
    """Context passed to the Evolution Engine when triggering evolution.

    Contains the information needed for the LLM analyzer to decide
    whether and how to evolve a Skill.
    """

    agent_id: str = Field(min_length=1)
    task_id: str = Field(min_length=1)
    task_description: str = ""
    task_completed: bool = False
    skill_ids_used: list[str] = Field(default_factory=list)
    skills_applied: list[str] = Field(default_factory=list)
    skills_fell_back: list[str] = Field(default_factory=list)
    execution_output: str | None = None
    execution_error: str | None = None

File: crates/ap-core/src/models/evolution.rs
//! Self-Evolution Engine models: SkillRecord, EvolutionType, SkillOrigin, SkillLineage.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::models::common::utc_now;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EvolutionType {
    Fix,
    Derived,
    Captured,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum SkillOrigin {
    #[default]
    Imported,
    Captured,
    Derived,
    Fixed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SkillLineage {
    #[serde(default)]
    pub origin: SkillOrigin,
    #[serde(default)]
    pub generation: u32,
    #[serde(default)]
    pub parent_skill_ids: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content_diff: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content_snapshot: Option<std::collections::HashMap<String, String>>,
}

impl Default for SkillLineage {
    fn default() -> Self {
        Self {
            origin: SkillOrigin::Imported,
            generation: 0,
            parent_skill_ids: Vec::new(),
            content_diff: None,
            content_snapshot: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SkillRecord {
    pub id: String,
    pub name: String,
    #[serde(default = "default_version")]
    pub version: String,
    #[serde(default)]
    pub lineage: SkillLineage,
    #[serde(default)]
    pub directory: String,
    #[serde(default = "default_true")]
    pub is_active: bool,
    #[serde(default)]
    pub total_selections: u64,
    #[serde(default)]
    pub total_applied: u64,
    #[serde(default)]
    pub total_completions: u64,
    #[serde(default)]
    pub total_fallbacks: u64,
    #[serde(default = "utc_now")]
    pub first_seen: DateTime<Utc>,
    #[serde(default = "utc_now")]
    pub last_updated: DateTime<Utc>,
}

fn default_version() -> String { "1.0.0".to_string() }
fn default_true() -> bool { true }

impl SkillRecord {
    /// Validate counter invariants.
    /// Python source: models/evolution.py `_validate_counters` — 5 checks.
    pub fn validate_counters(&self) -> Result<(), String> {
        if self.total_selections == 0
            && (self.total_applied != 0 || self.total_fallbacks != 0)
        {
            return Err("zero selections requires zero applied and zero fallbacks".into());
        }
        if self.total_applied > self.total_selections {
            return Err("total_applied cannot exceed total_selections".into());
        }
        if self.total_completions > self.total_applied {
            return Err("total_completions cannot exceed total_applied".into());
        }
        if self.total_fallbacks > self.total_applied {
            return Err("total_fallbacks cannot exceed total_applied".into());
        }
        if self.total_completions + self.total_fallbacks > self.total_applied {
            return Err("total_completions + total_fallbacks cannot exceed total_applied".into());
        }
        Ok(())
    }

    /// Shared counter validation logic used by both SkillRecord and EvolutionMetrics.
    pub fn validate_counters_from_parts(
        selections: u64,
        applied: u64,
        completions: u64,
        fallbacks: u64,
    ) -> Result<(), String> {
        if selections == 0 && (applied != 0 || fallbacks != 0) {
            return Err("zero selections requires zero applied and zero fallbacks".into());
        }
        if applied > selections {
            return Err("total_applied cannot exceed total_selections".into());
        }
        if completions > applied {
            return Err("total_completions cannot exceed total_applied".into());
        }
        if fallbacks > applied {
            return Err("total_fallbacks cannot exceed total_applied".into());
        }
        if completions + fallbacks > applied {
            return Err("total_completions + total_fallbacks cannot exceed total_applied".into());
        }
        Ok(())
    }
}

/// Standalone evolution metrics with same counter validators as SkillRecord.
///
/// Python source: models/evolution.py:97-123 `EvolutionMetrics`
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct EvolutionMetrics {
    #[serde(default)]
    pub total_selections: u64,
    #[serde(default)]
    pub total_applied: u64,
    #[serde(default)]
    pub total_completions: u64,
    #[serde(default)]
    pub total_fallbacks: u64,
}

impl EvolutionMetrics {
    pub fn validate(&self) -> Result<(), String> {
        SkillRecord::validate_counters_from_parts(
            self.total_selections, self.total_applied,
            self.total_completions, self.total_fallbacks,
        )
    }
}

/// Context passed to evolver with task/agent metadata.
///
/// Python source: models/evolution.py:126-141 `EvolutionContext`
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct EvolutionContext {
    #[serde(default)]
    pub agent_id: String,
    #[serde(default)]
    pub task_id: String,
    #[serde(default)]
    pub skill_ids_used: Vec<String>,
    #[serde(default)]
    pub task_description: String,
    #[serde(default)]
    pub task_result: Option<String>,
    #[serde(default)]
    pub error_info: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_record() -> SkillRecord {
        SkillRecord {
            id: "s1".into(),
            name: "test-skill".into(),
            version: default_version(),
            lineage: SkillLineage::default(),
            directory: String::new(),
            is_active: true,
            total_selections: 0,
            total_applied: 0,
            total_completions: 0,
            total_fallbacks: 0,
            first_seen: utc_now(),
            last_updated: utc_now(),
        }
    }

    #[test]
    fn valid_zero_counters() {
        assert!(make_record().validate_counters().is_ok());
    }

    #[test]
    fn invalid_applied_exceeds_selections() {
        let mut r = make_record();
        r.total_selections = 5;
        r.total_applied = 10;
        assert!(r.validate_counters().is_err());
    }

    #[test]
    fn valid_nonzero_counters() {
        let mut r = make_record();
        r.total_selections = 100;
        r.total_applied = 80;
        r.total_completions = 70;
        r.total_fallbacks = 10;
        assert!(r.validate_counters().is_ok());
    }

    #[test]
    fn evolution_metrics_validate_delegates() {
        let metrics = EvolutionMetrics {
            total_selections: 10,
            total_applied: 5,
            total_completions: 4,
            total_fallbacks: 1,
        };
        assert!(metrics.validate().is_ok());
    }

    #[test]
    fn evolution_metrics_invalid() {
        let metrics = EvolutionMetrics {
            total_selections: 5,
            total_applied: 10,
            total_completions: 0,
            total_fallbacks: 0,
        };
        assert!(metrics.validate().is_err());
    }
}
